# What is Pandas?

# Pandas is a Python library used for working with data sets.

# It has functions for analyzing, cleaning, exploring, and manipulating data.

# The name "Pandas" has a reference to both "Panel Data", and "Python Data Analysis" and was created by Wes McKinney in 2008.

# Why Use Pandas?
# Pandas allows us to analyze big data and make conclusions based on statistical theories.

# Pandas can clean messy data sets, and make them readable and relevant.

# Relevant data is very important in data science.

In [4]:
import pandas

mydataset = {
  'cars': ["BMW", "Volvo", "Ford"],
  'passings': [3, 7, 2]
}

myvar = pandas.DataFrame(mydataset)

print(myvar)

    trip_id        date  hour         zone  distance_km  duration_min    fare  waiting_time_min  passengers payment_type weather  driver_supply  demand
0     T0001  2026-08-11     9     Downtown         2.97          13.2   15.23               4.5           1         Card   Clear             92     109
1     T0002  2026-08-03    12   Industrial         8.10          18.5   24.62               5.6           1         Card   Clear             81      47
2     T0003  2026-08-02    13  Residential         5.97          16.8   20.06               3.4           2          App   Clear             74      74
3     T0004  2026-08-01    13  Residential         2.43          10.3    9.65               5.9           1         Card    Rain            107      93
4     T0005  2026-08-06    13   Industrial         7.73          15.3   26.19               5.0           2          App   Clear             52      65
5     T0006  2026-08-06     9     Downtown         3.58          12.0   16.90           

In [ ]:
import pandas as pd

mydataset = {
  'cars': ["BMW", "Volvo", "Ford"],
  'passings': [3, 7, 2]
}

myvar = pd.DataFrame(mydataset)

print(myvar)

# What is a Series?
# A Pandas Series is like a column in a table.

# It is a one-dimensional array holding data of any type.

In [ ]:
# Create a simple Pandas Series from a list:

import pandas as pd

a = [1, 7, 2]

myvar = pd.Series(a)

print(myvar)

In [5]:
# Create Labels
# With the index argument, you can name your own labels.

# Create your own labels:

import pandas as pd

a = [1, 7, 2]

myvar = pd.Series(a, index = ["x", "y", "z"])

print(myvar)

x    1
y    7
z    2
dtype: int64


In [6]:
# Key/Value Objects as Series
# You can also use a key/value object, like a dictionary, when creating a Series.

# Create a simple Pandas Series from a dictionary:

import pandas as pd

calories = {"day1": 420, "day2": 380, "day3": 390}

myvar = pd.Series(calories)

print(myvar)


day1    420
day2    380
day3    390
dtype: int64


In [7]:
# Create a Series using only data from "day1" and "day2":

import pandas as pd

calories = {"day1": 420, "day2": 380, "day3": 390}

myvar = pd.Series(calories, index = ["day1", "day2"])

print(myvar)

day1    420
day2    380
dtype: int64


# DataFrames
# Data sets in Pandas are usually multi-dimensional tables, called DataFrames.

# Series is like a column, a DataFrame is the whole table.

In [10]:
# Create a DataFrame from two Series:

import pandas as pd

data = {
  "calories": [420, 380, 390],
  "duration": [50, 40, 45]
}

myvar = pd.DataFrame(data)

print(myvar)

   calories  duration
0       420        50
1       380        40
2       390        45


In [13]:
import pandas as pd

data = {
  "calories": [420, 380, 390],
  "duration": [50, 40, 45]
}

#load data into a DataFrame object:
df = pd.DataFrame(data)

print(df) 

# Pandas use the loc attribute to return one or more specified row(s)

# Return row 0:

#refer to the row index:
print(df.loc[0])

   calories  duration
0       420        50
1       380        40
2       390        45
calories    420
duration     50
Name: 0, dtype: int64


In [14]:
#use a list of indexes:
print(df.loc[[0, 1]])

   calories  duration
0       420        50
1       380        40


In [15]:
# Named Indexes
# With the index argument, you can name your own indexes.

# Add a list of names to give each row a name:

import pandas as pd

data = {
  "calories": [420, 380, 390],
  "duration": [50, 40, 45]
}

df = pd.DataFrame(data, index = ["day1", "day2", "day3"])

print(df) 

      calories  duration
day1       420        50
day2       380        40
day3       390        45


In [16]:
# Locate Named Indexes
# Use the named index in the loc attribute to return the specified row(s).

#refer to the named index:
print(df.loc["day2"])

calories    380
duration     40
Name: day2, dtype: int64


# Load Files Into a DataFrame
### the first step for DS/DE

In [20]:
# Load a comma separated file (CSV file) into a DataFrame:

import pandas as pd

df = pd.read_csv('data.csv')

print(df) 

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
..        ...    ...       ...       ...
164        60    105       140     290.8
165        60    110       145     300.0
166        60    115       145     310.2
167        75    120       150     320.4
168        75    125       150     330.4

[169 rows x 4 columns]


# Read JSON
# Big data sets are often stored, or extracted as JSON.

# JSON is plain text, but has the format of an object, and is well known in the world of programming, including Pandas.

In [22]:
import pandas as pd

df = pd.read_json('data.json')

print(df.to_string()) 

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.5
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112       NaN
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45   

# Dictionary as JSON
### JSON objects have the same format as Python dictionaries.

In [24]:
# Load a Python Dictionary into a DataFrame:

import pandas as pd

data = {
  "Duration":{
    "0":60,
    "1":60,
    "2":60,
    "3":45,
    "4":45,
    "5":60
  },
  "Pulse":{
    "0":110,
    "1":117,
    "2":103,
    "3":109,
    "4":117,
    "5":102
  },
  "Maxpulse":{
    "0":130,
    "1":145,
    "2":135,
    "3":175,
    "4":148,
    "5":127
  },
  "Calories":{
    "0":409,
    "1":479,
    "2":340,
    "3":282,
    "4":406,
    "5":300
  }
}

df = pd.DataFrame(data)

print(df) 

   Duration  Pulse  Maxpulse  Calories
0        60    110       130       409
1        60    117       145       479
2        60    103       135       340
3        45    109       175       282
4        45    117       148       406
5        60    102       127       300


In [25]:
# Get a quick overview by printing the first 10 rows of the DataFrame:

import pandas as pd

df = pd.read_csv('data.csv')

print(df.head(10))

   Duration  Pulse  Maxpulse  Calories
0        60    110       130     409.1
1        60    117       145     479.0
2        60    103       135     340.0
3        45    109       175     282.4
4        45    117       148     406.0
5        60    102       127     300.0
6        60    110       136     374.0
7        45    104       134     253.3
8        30    109       133     195.1
9        60     98       124     269.0


In [26]:
# Print the last 5 rows of the DataFrame:

print(df.tail()) 

     Duration  Pulse  Maxpulse  Calories
164        60    105       140     290.8
165        60    110       145     300.0
166        60    115       145     310.2
167        75    120       150     320.4
168        75    125       150     330.4


In [27]:
# Print information about the data:

print(df.info()) 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 169 entries, 0 to 168
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Duration  169 non-null    int64  
 1   Pulse     169 non-null    int64  
 2   Maxpulse  169 non-null    int64  
 3   Calories  164 non-null    float64
dtypes: float64(1), int64(3)
memory usage: 5.4 KB
None


### The info() method also tells us how many Non-Null values there are present in each column, 
### and in our data set it seems like there are 164 of 169 Non-Null values in the "Calories" column.

### Which means that there are 5 rows with no value at all, in the "Calories" column, for whatever reason.

### Empty values, or Null values, can be bad when analyzing data, and you should consider removing rows with empty values. 
### This is a step towards what is called cleaning data。

### Data cleaning means fixing bad data in your data set.

### Bad data could be:

### Empty cells
### Data in wrong format
### Wrong data
### Duplicates

### Remove Rows
### One way to deal with empty cells is to remove rows that contain empty cells.

### This is usually OK, since data sets can be very big, and removing a few rows will not have a big impact on the result.

In [34]:
# Return a new Data Frame with no empty cells:

import pandas as pd

df = pd.read_csv('data_1.csv')

new_df = df.dropna()

print(new_df.to_string())
len(new_df)

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45    105       132     246.0
24         60   

164

In [31]:
# By default, the dropna() method returns a new DataFrame, and will not change the original.

# If you want to change the original DataFrame, use the inplace = True argument:

# Remove all rows with NULL values:

import pandas as pd

df = pd.read_csv('data_1.csv')

df.dropna(inplace = True)

print(df.to_string())

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45    105       132     246.0
24         60   

In [37]:
# Replace Empty Values
# Another way of dealing with empty cells is to insert a new value instead.

# This way you do not have to delete entire rows just because of some empty cells.

# The fillna() method allows us to replace empty cells with a value:

# Replace NULL values with the number 130:

import pandas as pd

df = pd.read_csv('data_1.csv')

df.fillna(130, inplace = True)


In [ ]:
# Replace NULL values in the "Calories" columns with the number 130:

import pandas as pd

df = pd.read_csv('data_1.csv')

df.fillna({"Calories": 130}, inplace=True)

# Replace Using Mean, Median, or Mode
# A common way to replace empty cells, is to calculate the mean, median or mode value of the column.

# Pandas uses the mean() median() and mode() methods to calculate the respective values for a specified column:

In [38]:
# Calculate the MEAN, and replace any empty values with it:
# Mean = the average value (the sum of all values divided by number of values).

import pandas as pd

df = pd.read_csv('data_1.csv')

x = df["Calories"].mean()

df.fillna({"Calories": x}, inplace=True)

In [39]:
# Calculate the MEDIAN, and replace any empty values with it:
# Median = the value in the middle, after you have sorted all values ascending.

import pandas as pd

df = pd.read_csv('data_1.csv')

x = df["Calories"].median()

df.fillna({"Calories": x}, inplace=True)

In [40]:
# Calculate the MODE, and replace any empty values with it:
# Mode = the value that appears most frequently.

import pandas as pd

df = pd.read_csv('data_1.csv')

x = df["Calories"].mode()[0]

df.fillna({"Calories": x}, inplace=True)

# Pandas - Cleaning Data of Wrong Format

### Data of Wrong Format

### Cells with data of wrong format can make it difficult, or even impossible, to analyze data.

### To fix it, you have two options: remove the rows, or convert all cells in the columns into the same format.

In [42]:
# convert all cells in the 'Date' column into dates. Pandas has a to_datetime() method for this:

import pandas as pd

df = pd.read_csv('data_1.csv')

df['Date'] = pd.to_datetime(df['Date'], format='mixed')

print(df.to_string())

    Duration       Date  Pulse  Maxpulse  Calories
0         60 2020-12-01    110       130     409.1
1         60 2020-12-02    117       145     479.0
2         60 2020-12-03    103       135     340.0
3         45 2020-12-04    109       175     282.4
4         45 2020-12-05    117       148     406.0
5         60 2020-12-06    102       127     300.0
6         60 2020-12-07    110       136     374.0
7        450 2020-12-08    104       134     253.3
8         30 2020-12-09    109       133     195.1
9         60 2020-12-10     98       124     269.0
10        60 2020-12-11    103       147     329.3
11        60 2020-12-12    100       120     250.7
12        60 2020-12-12    100       120     250.7
13        60 2020-12-13    106       128     345.3
14        60 2020-12-14    104       132     379.3
15        60 2020-12-15     98       123     275.0
16        60 2020-12-16     98       120     215.2
17        60 2020-12-17    100       120     300.0
18        45 2020-12-18     90 

In [43]:
# Removing Rows
# The result from the converting in the example above gave us a NaT value, which can be handled as a NULL value, 
# and we can remove the row by using the dropna() method.

# Remove rows with a NULL value in the "Date" column:

df.dropna(subset=['Date'], inplace = True)

# Pandas - Fixing Wrong Data

In [ ]:
# Set "Duration" = 45 in row 7:

df.loc[7, 'Duration'] = 45

In [ ]:
# Loop through all values in the "Duration" column.

# If the value is higher than 120, set it to 120:

for x in df.index:
  if df.loc[x, "Duration"] > 120:
    df.loc[x, "Duration"] = 120

In [ ]:
# Delete rows where "Duration" is higher than 120:

for x in df.index:
  if df.loc[x, "Duration"] > 120:
    df.drop(x, inplace = True)

# Pandas - Removing Duplicates

In [44]:
# Duplicate rows are rows that have been registered more than one time.
# To discover duplicates, we can use the duplicated() method.

# The duplicated() method returns a Boolean values for each row:
    
# Returns True for every row that is a duplicate, otherwise False:

print(df.duplicated())

0     False
1     False
2     False
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11    False
12     True
13    False
14    False
15    False
16    False
17    False
18    False
19    False
20    False
21    False
23    False
24    False
25    False
26    False
27    False
28    False
29    False
30    False
31    False
dtype: bool


In [ ]:
# Remove all duplicates:

df.drop_duplicates(inplace = True)

# Remember: The (inplace = True) will make sure that the method does NOT return a new DataFrame, 
# but it will remove all duplicates from the original DataFrame.